include/DeterministicNoiseSimulator.hpp

```diff
@@ -49,31 +49,29 @@ class DeterministicNoiseSimulator: public Simulator<Config> {
     [[nodiscard]] std::size_t getActiveNodeCount() const override { return Simulator<Config>::dd->dUniqueTable.getActiveNodeCount(); }
     [[nodiscard]] std::size_t getMaxNodeCount() const override { return Simulator<Config>::dd->dUniqueTable.getMaxActiveNodes(); }
 
-    [[nodiscard]] std::size_t countNodesFromRoot() {
-        size_t tmp;
+    [[nodiscard]] std::size_t countNodesFromRoot() override {
         if (useDensityMatrixType) {
             qc::DensityMatrixDD::alignDensityEdge(rootEdge);
-            tmp = Simulator<Config>::dd->size(rootEdge);
+            const std::size_t tmp = Simulator<Config>::dd->size(rootEdge);
             qc::DensityMatrixDD::setDensityMatrixTrue(rootEdge);
-        } else {
-            tmp = Simulator<Config>::dd->size(rootEdge);
+            return tmp;
         }
-        return tmp;
+        return Simulator<Config>::dd->size(rootEdge);
     }
 
     qc::DensityMatrixDD rootEdge{};
 
 private:
-    const std::unique_ptr<qc::QuantumComputation>& qc;
-    const std::vector<dd::NoiseOperations>         noiseEffects;
+    const std::unique_ptr<qc::QuantumComputation>& qc; // NOLINT(cppcoreguidelines-avoid-const-or-ref-data-members)
```

Comment:
Good catch, moving into a plain `std::unique_ptr` is enough.

Reply:
To be honest, seeing the changes you made just now makes me thing whether a `std::shared_ptr` would actually be more appropriate. We don't really want the simulator to take ownership of the circuit. The circuit should remain useable elsewhere. Especially since we offer no way to transfer ownership back from the simulator.

Reply:
Agreed.
I think that's how we deal with that in QCEC at the moment in order to apply optimizations to the circuit without changing the original circuits.

I'd say this should be good to go then, right? Or would you want to revert the latest round of changes and do it "properly" in a separate PR?

Reply:
The complaining was about something in tests (bit obscure though). 

I'd change it to moving as in the other simulators. Thinking about it,  the ref to a unique_ptr seems like just a bad way to do what a shared pointer does, doesn't it?

Reply:
Yeah. I think the `const std::unique_ptr&` was a bad design on our end and should have been a `std::shared_ptr`.
Since we do not manipulate the circuits at the moment in the simulator, this would be a fair solution for now.
I do agree with you though, that it might make sense to alter the circuit in the simulator (e.g., apply some optimizations). And at that point, we almost certainly need a copy.

Reply:
I feel like the simulator taking ownership of the quantum simulator is the better solution compared to using a shared pointer. So I'd keep it as is at the moment. Having the quantum computation in a shared pointer has way more opportunity for errors to go undetected, where as there is at least a warning if you are using a moved-from object.

Ideally we think about how we can have easy and cheap copying for quantum computations so we can pass them by value.

Reply:
Just to be clear: Before this PR, the circuit was not owned by the simulator and that newer caused problems until now. With the shared pointer approach, the outcome would just be the same as before this PR, but in a cleaner fashion (e.g., the copy constructor of the Simulator class would not get implicitly deleted). With the unique pointer approach (as it is right now in the PR) you introduce a breaking change (ownership of the circuit is lost), without gaining much (copy constructor, e.g., is still implicitly deleted).

BTW: Quantum circuits can already be copied fairly easily by calling `qc.clone()`.

Reply:
Before this PR, the `DeterministicNoiseSimulator` and `StochasticNoiseSimulator` did not take ownership. The `CircuitSimulator` (and its derived classes `HybridSchrodingerFeynmanSimulator`, `PathSimulator`, and `UnitarySimulator`) did take the ownership before this PR. Given that we are not exposing these the noisy simulator via Python, I'd rather have a breaking change in them and unify the behaviour regarding taking ownership for all simulators.

I'd also favor using `QuantumComputation::clone` as a means to create a temporary to be moved into the simulator.

Reply:
Didn't think of that and agree with your reasoning. Unified behavior across all simulators certainly is the way to go.
We can change that unified behavior in a subsequent PR.

Just one last comment and then I'll shut up: why is the additional member in the noise-aware simulators necessary at all? couldn't those just inherit from the CircuitSimulator? 🤷🏼

Reply:
I'll leave it as it is now and save the ideas in #201 :+1: